Add Support for Linux

Tests are currently failing on Linux, although it's unclear exactly why. Perhaps something to do with a behavioral difference in the Swift REPL between macOS and Linux.

👋 I just tried setting up a PR, and I think I'm running headfirst into this. The error I'm getting on `swift test --enable-test-discovery` (using the swift:5.2 ubuntu linux docker image) is:

```
Test Suite 'All tests' started at 2020-04-20 04:42:02.871
Test Suite 'debug.xctest' started at 2020-04-20 04:42:02.874
Test Suite 'DocTestTests' started at 2020-04-20 04:42:02.874
Test Case 'DocTestTests.testRunner' started at 2020-04-20 04:42:02.876
/DocTest/Tests/DocTestTests/DocTestTests.swift:20: error: DocTestTests.testRunner : XCTAssertEqual failed: ("0") is not equal to ("3") -
Fatal error: Index out of range: file /home/buildnode/jenkins/workspace/oss-swift-5.2-package-linux-ubuntu-18_04/swift/stdlib/public/core/ContiguousArrayBuffer.swift, line 444
Current stack trace:
0    libswiftCore.so                    0x00007fa668364f90 swift_reportError + 50
1    libswiftCore.so                    0x00007fa6683d6a60 _swift_stdlib_reportFatalErrorInFile + 115
2    libswiftCore.so                    0x00007fa6680eb735 <unavailable> + 1382197
3    libswiftCore.so                    0x00007fa6680eb257 <unavailable> + 1380951
4    libswiftCore.so                    0x00007fa6680eaf23 <unavailable> + 1380131
5    libswiftCore.so                    0x00007fa6680ea890 _assertionFailure(_:_:file:line:flags:) + 525
6    libswiftCore.so                    0x00007fa6680d0ed0 Array.subscript.getter + 123
7    DocTestPackageTests.xctest         0x0000563b71165efc <unavailable> + 761596
8    DocTestPackageTests.xctest         0x0000563b71166c9f <unavailable> + 765087
9    libXCTest.so                       0x00007fa66863a730 XCTAssertTrue(_:_:file:line:) + 38
10   DocTestPackageTests.xctest         0x0000563b71165ba5 <unavailable> + 760741
11   DocTestPackageTests.xctest         0x0000563b7115e6ff <unavailable> + 730879
12   DocTestPackageTests.xctest         0x0000563b711655b1 <unavailable> + 759217
13   DocTestPackageTests.xctest         0x0000563b71165229 <unavailable> + 758313
14   DocTestPackageTests.xctest         0x0000563b7116500f <unavailable> + 757775
15   DocTestPackageTests.xctest         0x0000563b71165214 <unavailable> + 758292
16   libXCTest.so                       0x00007fa668633fd1 <unavailable> + 200657
17   libXCTest.so                       0x00007fa668633dde <unavailable> + 200158
18   libXCTest.so                       0x00007fa668633d64 <unavailable> + 200036
19   libXCTest.so                       0x00007fa6686340a9 <unavailable> + 200873
20   libXCTest.so                       0x00007fa6686254e7 <unavailable> + 140519
21   libXCTest.so                       0x00007fa668631ef0 XCTestCase.invokeTest() + 55
22   libXCTest.so                       0x00007fa668631bc0 XCTestCase.perform(_:) + 157
23   libXCTest.so                       0x00007fa668636530 XCTest.run() + 150
24   libXCTest.so                       0x00007fa6686342e0 XCTestSuite.perform(_:) + 160
25   libXCTest.so                       0x00007fa668636530 XCTest.run() + 150
26   libXCTest.so                       0x00007fa6686342e0 XCTestSuite.perform(_:) + 160
27   libXCTest.so                       0x00007fa668636530 XCTest.run() + 150
28   libXCTest.so                       0x00007fa6686342e0 XCTestSuite.perform(_:) + 160
29   libXCTest.so                       0x00007fa668636530 XCTest.run() + 150
30   libXCTest.so                       0x00007fa66862fda0 XCTMain(_:) + 1524
31   DocTestPackageTests.xctest         0x0000563b71165301 <unavailable> + 758529
32   libc.so.6                          0x00007fa6659c1ab0 __libc_start_main + 231
33   DocTestPackageTests.xctest         0x0000563b71129a6a <unavailable> + 514666
Exited with signal code 4
```

I have a sneaky suspicion there's some sort of thread race scenario happening here, but I'm no expert.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -18,12 +18,14 @@ jobs:
         env:
           DEVELOPER_DIR: /Applications/Xcode_11.4.app/Contents/Developer
 
-  # linux:
-  #   runs-on: ubuntu-latest
-  #   container:
-  #     image: swift:5.2
-  #   steps:
-  #     - name: Checkout
-  #       uses: actions/checkout@v1
-  #     - name: Build and Test
-  #       run: swift test --enable-test-discovery
+  linux:
+    runs-on: ubuntu-latest
+
+    container:
+      image: swift:5.2
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v1
+      - name: Build and Test
+        run: swift test --enable-test-discovery

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -23,7 +23,7 @@ jobs:
 
     container:
       image: swift:5.2
-      options: --privileged --cap-add sys_ptrace
+      options: --cap-add=SYS_PTRACE --security-opt seccomp=unconfined --security-opt apparmor=unconfined
 
     steps:
       - name: Checkout

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -23,6 +23,7 @@ jobs:
 
     container:
       image: swift:5.2
+      options: --privileged --cap-add sys_ptrace
 
     steps:
       - name: Checkout

--- a/Changelog.md
+++ b/Changelog.md
@@ -7,6 +7,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- Added documentation and testing infrastructure for Linux support.
+  #17 by @heckj and @mattt.
+
 ## [0.0.4] - 2020-04-26
 
 ### Changed

--- a/README.md
+++ b/README.md
@@ -29,7 +29,21 @@ just like a unit test.
 
 ## Requirements
 
-- macOS with Swift 5.2 / Xcode 11.4
+- Swift 5.2
+
+> **Note**:
+> When running `swift-doctest` from a Docker container,
+> pass the following options:
+>
+> - `--cap-add=SYS_PTRACE`  
+>   _(allows the launched process to trace system calls from)_
+> - `--security-opt seccomp=unconfined`  
+>   _(runs the container without the [default security profile][seccomp])_
+> - `--security-opt apparmor=unconfined`  
+>   _(runs the container without the [AppArmor security profile][apparmor])_
+> 
+> Without these permissions,
+> DocTest cannot launch the Swift REPL to evaluate code statements.
 
 ## Installation
 
@@ -66,7 +80,7 @@ OPTIONS:
                           The path to the swift executable. (default:
                           /usr/bin/swift)
   -p, --package           Whether to run the REPL through Swift Package Manager
-                          (`swift run --repl`). 
+                          (`swift run --repl`).
   --assumed-filename <assumed-filename>
                           The assumed filename to use for reporting when
                           parsing from standard input. (default: Untitled.swift)
@@ -193,3 +207,6 @@ MIT
 ## Contact
 
 Mattt ([@mattt](https://twitter.com/mattt))
+
+[seccomp]: https://docs.docker.com/engine/security/seccomp/
+[apparmor]: https://docs.docker.com/engine/security/apparmor/

